sdks/java/extensions/protobuf/src/test/java/org/apache/beam/sdk/extensions/protobuf/ProtoSchemaValuesTest.java

```diff
@@ -0,0 +1,670 @@
+              .build(),
+          change(COMPLEX_DEFAULT_ROW, "z", mapRow)
+        });
+    data.add(
+        new Object[] {
+          "subRow",
+          Proto3SchemaMessages.Message.newBuilder()
+              .setMessage(
+                  Proto3SchemaMessages.Primitive.newBuilder()
+                      .setPrimitiveString("we love strings")
+                      .build())
+              .build(),
+          Row.withSchema(MESSAGE_SCHEMA)
+              .addValue(change(PRIMITIVE_DEFAULT_ROW, "primitive_string", "we love strings"))
+              .addValue(null)
+              .build()
+        });
+    data.add(
+        new Object[] {
+          "subRow+subArrayOfRow",
+          Proto3SchemaMessages.Message.newBuilder()
+              .setMessage(Proto3SchemaMessages.Primitive.newBuilder().setPrimitiveInt32(42).build())
+              .addRepeatedMessage(
+                  Proto3SchemaMessages.Primitive.newBuilder().setPrimitiveInt32(69).build())
+              .addRepeatedMessage(
+                  Proto3SchemaMessages.Primitive.newBuilder().setPrimitiveInt32(70).build())
+              .addRepeatedMessage(
+                  Proto3SchemaMessages.Primitive.newBuilder().setPrimitiveInt32(71).build())
+              .build(),
+          Row.withSchema(MESSAGE_SCHEMA)
+              .addValue(change(PRIMITIVE_DEFAULT_ROW, "primitive_int32", 42))
+              .addArray(
+                  change(PRIMITIVE_DEFAULT_ROW, "primitive_int32", 69),
+                  change(PRIMITIVE_DEFAULT_ROW, "primitive_int32", 70),
+                  change(PRIMITIVE_DEFAULT_ROW, "primitive_int32", 71))
+              .build()
+        });
+    return data;
+  }
+
+  private static Row change(Row row, Object field, Object value) {
+    int index = -1;
+    List<Schema.Field> fields = row.getSchema().getFields();
+    for (int i = 0; i < fields.size(); i++) {
+      Schema.Field f = fields.get(i);
+      if (f.getName().equals(field)) {
+        index = i;
+        break;
+      }
+    }
+
+    Object[] objects = row.getValues().toArray();
+    objects[index] = value;
+    return Row.withSchema(row.getSchema()).addValues(objects).build();
+  }
+
+  private void setup() {
+    ProtoSchemaProvider protoSchemaProvider = new ProtoSchemaProvider();
+    TypeDescriptor typeDescriptor = TypeDescriptor.of(this.proto.getClass());
+
+    toRowFunction = protoSchemaProvider.toRowFunction(typeDescriptor);
+    fromRowFunction = protoSchemaProvider.fromRowFunction(typeDescriptor);
+  }
+
+  private void setupForDynamicMessage() {
+    ProtoDomain domain = ProtoDomain.buildFrom(proto.getDescriptorForType());
+    ProtoSchema protoSchema =
+        ProtoSchema.newBuilder(domain).forDescriptor(proto.getDescriptorForType());
+
+    toRowFunction = protoSchema.getSchemaCoder().getToRowFunction();
+    fromRowFunction = protoSchema.getSchemaCoder().getFromRowFunction();
+  }
+
+  @Test
+  public void testRowAndBack() {
+    setup();
```

Comment:
I wonder if this could be re-structured a bit so that you could use `@Before` instead of calling the setup functions yourself. Maybe if you have two different Parameterized test classes that draw from the same data? One for message and one for dynamic message. You can have multiple suites in one outer class if you annotate it with `@RunWith(Enclosed.class)` (like I did [here](https://github.com/apache/beam/pull/9446/files#diff-cd81b3c0b9fb7c9081a9be718cc5cdd1R77))

This is definitely a nitpicky thing... totally fine to leave it as is.